src/cutlass_mlp.cu

```diff
@@ -314,6 +327,606 @@ void CutlassMLP<T>::backward_impl(
+			output_weight_matrix(use_inference_params), // weights
+			fc_last_output, // p
+			m_output_activation == Activation::None ? dL_doutput : backward_output_tmp, // dL_doutput
+			dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2_ddL1dinput of current layer
+			ddoutputdp_dp.at(m_n_hidden_matmuls+1), // ddoutputdp_dp
+			dL2dinput.at(m_n_hidden_matmuls+1), // dL2dinput of current layer
+			param_gradients_mode
+		);		
+
+		if (dL_ddLdoutput) { // if dL_ddLdoutput is not nullptr
+			bool return_output_dL2_ddL1doutput = compute_dL2_ddL1doutput<FullLayer, T>(
+				stream,
+				false, // is_inference
+				m_output_activation,
+				output_weight_matrix(use_inference_params),
+				fc_last_output, // p
+				dL2_ddL1doutput.at(bwd_bwd_idx-1), // dL2d_dL1dinput
+				*dL_ddLdoutput, // dL2_ddL1doutput
+				param_gradients_mode
+			);
+		}
+	}
+	
+	// 1st order backward of dL2dinput and dL2dw in reverse order
+	std::vector<SyncedMultiStream> multi_streams_1st;
+	multi_streams_1st.emplace_back(stream, 2);
+	int split_k_factor = batch_size / std::min((uint32_t)(1 << 12), batch_size);
+	const float param_gradient_beta = 1.0f; // param_gradients_mode == GradientMode::Accumulate
+
+	for (uint32_t i = m_n_hidden_layers; i > 0; i--) {
+
+		GPUMatrixDynamic<T> dL2dp = GPUMatrix<T>{fc_output.at(i-1).rows(), fc_output.at(i-1).cols(), stream};
+		activation_backward_gpu(stream, dL2dp.n_elements(), m_activation, fc_output.at(i-1).data(), dL2dinput.at(i).data(), dL2dp.data());
+
+		// compute 1st order dL2dw
+		if (param_gradients_mode != GradientMode::Ignore) {
+			if (i - 1) {
+				fc_multiply_split_k<FullLayerK>(multi_streams_1st.back().get(1), dL2dp, forward.hidden.at(i-2).transposed(), gradient_matrix_at(i-2), split_k_factor, param_gradient_beta);
+			} else { // i - 1 == 0: input layer
+				fc_multiply_split_k<FullLayerK>(multi_streams_1st.back().get(1), dL2dp, input.transposed(), input_gradient_matrix(), split_k_factor, param_gradient_beta);
+			}
+		}
+		
+		if (i - 1) {
+			GPUMatrixDynamic<T> dL_dinput_1st_order = GPUMatrix<T>{dL2dinput.at(i-1).rows(), dL2dinput.at(i-1).cols(), stream};
+			// weight_matrix_at(use_inference_params, i-2).transposed().layout() == CM
+			fc_multiply<FullLayer>(stream, weight_matrix_at(use_inference_params, i-2).transposed().cm(), dL2dp, dL_dinput_1st_order);
+			linear_kernel(element_wise_add<T>, 0, stream, dL2dinput.at(i-1).n() * dL2dinput.at(i-1).m(), dL_dinput_1st_order.data(), dL2dinput.at(i-1).data());
+		} else if (i - 1 == 0 && dL_dinput) { // input layer
+			GPUMatrixDynamic<T> dL_dinput_1st_order = GPUMatrix<T>{dL2dinput.at(i-1).rows(), dL2dinput.at(i-1).cols(), stream};
+			// input_weight_matrix(use_inference_params).transposed().layout() == CM
+			fc_multiply<FullLayer>(stream, input_weight_matrix(use_inference_params).transposed().cm(), dL2dp, dL_dinput_1st_order);
+			linear_kernel(element_wise_add<T>, 0, stream, dL2dinput.at(i-1).n() * dL2dinput.at(i-1).m(), dL_dinput_1st_order.data(), dL2dinput.at(i-1).data());
```

Comment:
Use `matrix.n_elements()` instead of `matrix.n() * matrix.m()` to avoid looking up the same matrix multiple times.

(I'm not repeating this comment in the other places where this occurs; please also check those.)